### PR TITLE
Added experimental swarm deployment

### PR DESCRIPTION
**Pull request by krleh**
Added experimental support for docker swarm:

- Documentation: `docs/swarm_deployment.md`
- Stack template: `make/harbor-stack.tpl`
- Modifications to the build process

Related issues:
- #2868 (#3309, #2948, #2995)

### CODE REVIEW

File: docs/swarm_deployment.md [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+### Configuration
+
+Change your settings in `harbor.cfg` as usual. For the deployment with swarm the configuration has to be prepared using the `./prepare` script and the following parameters:
+
+```
+./prepare --experimental-swarm
+```
+
+### clair
```
Comment (clouderati) [inline docs/swarm_deployment.md]: s/clair/Clair

File: docs/swarm_deployment.md [hunk 2]
```diff
@@ -0,0 +1,56 @@
+**IMPORTANT** This deployment method is experimental, Docker Swarm experience is still required for volumes and container placement constraints!
+
+## Docker Swarm
+
+This document describes how to deploy **harbor** with **Docker Swarm**.
```
Comment (clouderati) [inline docs/swarm_deployment.md]: s/harbor/Harbor

File: docs/swarm_deployment.md [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+Change your settings in `harbor.cfg` as usual. For the deployment with swarm the configuration has to be prepared using the `./prepare` script and the following parameters:
+
+```
+./prepare --experimental-swarm
+```
+
+### clair
+
+When using **clair**  comment the according lines in `harbor-stack.yml` and enable **clair**:
```
Comment (clouderati) [inline docs/swarm_deployment.md]: s/clair/Clair/g

File: docs/swarm_deployment.md [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+```
+protocol=https
+ssl_cert = ./data/cert/server.crt
+ssl_cert_key = ./data/cert/server.key
+```
+
+Then place your cert and key files in `./data/cert/`.
+
+#### traefik
```
Comment (clouderati) [inline docs/swarm_deployment.md]: s/traefik/Traefik

File: docs/swarm_deployment.md [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+If you are using SSL with **traefik**, you have to change the realm property in `./common/config/registry/config.yml`: Go to `auth` -> `token` -> `realm` and replace "http" with "https". See: https://github.com/goharbor/harbor/issues/1097
```
Comment (clouderati) [inline docs/swarm_deployment.md]: s/traefik/Traefik

File: docs/swarm_deployment.md [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+Afterwards remove the `nginx`-service section in `./harbor-stack.yml` and add your traefik labels to the `portal`-service
+
+
+### Deploy stack
+**IMPORTANT** The stack configuration in `harbor-stack.yml` has to be modified manually to match your swarm environment:
+
+* **Take care of the volume locations!**
+* **Modify the placement constraints!**
+
+Afterwards deploy **harbor** to you swarm:
```
Comment (clouderati) [inline docs/swarm_deployment.md]: s/harbor/Harbor

File: docs/swarm_deployment.md [hunk 7]
```diff
@@ -0,0 +1,56 @@
+**IMPORTANT** This deployment method is experimental, Docker Swarm experience is still required for volumes and container placement constraints!
```
Comment (clouderati) [inline docs/swarm_deployment.md]: Sorry to be pedantic; can we please say it's "very" experimental? :smile: Just don't want anyone to try this early on and get frustrated by a snag.

s/experience/expertise

**Comment by krleh** [comment]
Please go ahead testing it, if you have any questions contact me on slack :+1:

**Comment by clouderati** [comment]
@krleh – my deepest apologies. I'll try to test this tomorrow and get back to you.

**Comment by clouderati** [comment]
Hi @krleh – sorry for the delay in getting back to you. Spoke to my fellow maintainers and the consensus is to hold off on merging Swarm support primarily because we're not in a position to support it. Our focus atm is `docker-compose` and Helm support.

Happy to link to a repo in your GitHub org from our docs if you'd like, though.

Apologies again for taking so long to get back to you.

**Comment by darkl0rd** [comment]
Now that docker-compose and helm support are stable and mature; can you please re-consider adding official swarm support?

**Comment by bjethwan** [comment]
Do we need to log a new feature request or issue?

**Comment by mmgil** [comment]
Swarm Please

**Pull request closed without merging**